spec.md

```diff
@@ -1617,6 +1642,53 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 |-----------|-----------|-------------|-------------------|
 | Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
 
+
+#### `ControllerModifyVolume`
+
+A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability.
+This RPC allows the CO to change mutable key attributes of a volume. 
+
+This operation MUST be idempotent. 
+The new mutable parameters in ControllerModifyVolume can be different from the existing mutable parameters. 
+
+
+```protobuf
+message ControllerModifyVolumeRequest {
+  option (alpha_message) = true;
+
+  // Contains identity information for the existing volume.
+  // This field is REQUIRED.
+  string volume_id = 1;
+
+  // Secrets required by plugin to complete modify volume request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 2 [(csi_secret) = true];
+
+  // Plugin specific parameters to apply, passed in as opaque key-value
+  // pairs. This field is OPTIONAL. The Plugin is responsible for
+  // parsing and validating these parameters. COs will treat these
+  // as opaque. COs MUST specify the intended value of every mutable
+  // parameter. Absent keys that were previously assigned a value MUST
+  // be interpreted as a deletion of the key from the set of mutable
+  // parameters.
+  map<string, string> mutable_parameters = 3;
```

Comment:
> Absent keys that were previously assigned a value MUST
  // be interpreted as a deletion of the key from the set of mutable
  // parameters.

Does this mean that calling `ControllerModifyVolume` without specifying anything in `mutable_parameters` will mean, user is trying to reset these `mutable_parameters` to whatever default values driver/provider supports? What if storage provider does not support going back to default values?

It also feels like, this would generally mean driver should keep track of each `mutable_parameter` it supports and reset them if the keys are unspecified?  I am not sure, if this will be possible for each driver.

Reply:
This wording was aimed mostly at the CO. I wanted to make it clear that if there are 2 mutable parameters "foo" and "bar" which were previously set, and the CO wants to modify "foo", it cannot just send a modify with the new value for "foo" and omit "bar", it needs to send the whole updated map of mutable parameters.

The goal here is to make it clear that the CO must implement a simple reconciler loop that always passes the whole map any time the map changes, and not something more complex where the CO computes a diff of what changed and sends only the diff.

Reply:
From k8s perspective since values in `ControllerModifyVolume` will be coming from respective VAC class, this means that, if k8s admin wants a volume to be associated with new VAC class, they must specify all parameters from previous VAC class or CO should compute `parameters`  of `ControllerModifyVolume` by using parameters from both old and new VAC class.

Another problem I have is with wording `MUST be interpreted as deletion of key`. I don't know how driver will interpret this and what they are supposed to do.

Reply:
If you have a better recommended wording I'm flexible here. I just recognized that in the previous spec wording that if you wanted to change two different values, you might have thought that you could do it as two separate modify calls each with one key in the map. I wanted to make it clear that the CO is supposed to send everything it knows each time, and not attempt to modify parameters one at a time.

I agree that it's not obvious what it means to delete/unset a value, and mandating that the SP does so not my intent. My point was that the CO should not omit values that it knows but does not intend to modify.

Reply:
> From k8s perspective since values in `ControllerModifyVolume` will be coming from respective VAC class, this means that, if k8s admin wants a volume to be associated with new VAC class, they must specify all parameters from previous VAC class or CO should compute `parameters` of `ControllerModifyVolume` by using parameters from both old and new VAC class.

My intent is that if the VAC changes, the reconciler would just send the parameters for the new VAC. The contents of the old VAC would be irrelevant, so neither the CO nor the SP would need to know them in order for modify to complete successfully.

Reply:
> My intent is that if the VAC changes, the reconciler would just send the parameters for the new VAC. The contents of the old VAC would be irrelevant, so neither the CO nor the SP would need to know them in order for modify to complete successfully.

It sounds like, what we want is to apply whatever parameter admin has provided in the VAC class and leave other parameters untouched. For example:

1. Volume gets created with no VAC class and lets say gets default iops of `10000` and throughput of `30`.
2. Admin associates the volume with VAC class `silver` , which has following properties:

```
iops: 20,0000
throughput: 100
```

So volume is modified accordingly.

3. Admin associates the volume with `gold` VAC class but somehow forgets to specify `throughput` parameter in VAC class. So `gold` VAC class has following properites:

```
iops: 75,0000
```

Now CO calls `ControllerModifyVolume` with `iops: 75,000` but this call does not reset the `throughput` to original value of `30` but leaves it at whatever last value was (which would be `100` from being attached to `silver` class).

If ^ flow is accurate then wording - `Absent keys that were previously assigned MUST be interpreted as a deletion` is problematic, because that would mean each VAC class should contain exhaustive list of parameters otherwise driver MUST somehow reset the original value of missing parameter.

If the intention is to apply parameters in `ControllerModifyVolume`  and ignore missing keys, then I am not sure why is `mutable_parameters` an `OPTIONAL` parameter in this call. It should not be optional and should not be empty.

But if the intention is indeed to reset the value of missing key to original value, then we have bunch of problems:

1.  This creates a error prone problem where each VAC admin creates MUST contain exhaustive list of key/value parameters, since any missing key will be interpreted as deletion and value has to be reset.
2. It creates a problem for driver authors, where they have to figure out - how to reset the value of missing key while  updating other parameters.

I would personally prefer to remove the wording about absent keys etc and make `mutable_parameters` non-optional and non-empty in `ControllerModifyVolume` RPC call. And add something like:

```
Plugin specific parameters to apply, passed in as opaque key-value
pairs. This field is REQUIRED and MUST not be empty. The Plugin is
responsible for parsing and validating these parameters. COs will
treat these as opaque. 
```

Reply:
So the map always needs to optional for the reason that non-mutable parameters is optional -- because the map might be empty!

But I agree that the language around deletion might be too restrictive. I just didn't want to have a situation where CO authors thought they had the freedom to modify keys one at time when the user had requested multiple changes (i.e. change from one VAC to another). I want to communicate that the CO should send all of the changes it knows about in a single call, and I was looking for words that mean that.

Reply:
So a `ControllerModifyVolume` call without `mutable_parameters` is a NO-OP right?

Reply:
> So a `ControllerModifyVolume` call without `mutable_parameters` is a NO-OP right?

I don't think we can make that blanket statement. Some SPs will have mutable parameters that needs a mechanism to clear or reset them. If we define "key in map is empty during modify call" to mean no-op, then it become impossible to clear or reset one of these parameters.

I am okay with SP's choosing to interpret an omitted key as no change for SOME keys on SOME SPs, but we also need to allow a mechanism to reliably clear or reset mutable settings, and my proposal is to require the CO to always pass every parameter it knows, and to leave it up to the SP whether to interpret absence as a no-op or a clear. This makes the CO's job very easy, and gives the SP the power to implement a variety of things, with the burden of documentation falling on the specific vendor.

Reply:
> to leave it up to the SP whether to interpret absence as a no-op or a clear.

That doesn't seem very consistent from API perspective. What is the use case of resetting the parameters? Can users not apply a different VAC to change or go back to previously used `mutable_parameters` ? At least in k8s APIs I think discussion is leading towards, not letting users set VAC back to nil once it has been set.

Reply:
Also btw, the CSI spec has following wording about maps:

> A repeated or map field listed as REQUIRED MUST contain at least 1 element.

So we do seem to have some provision for making maps as REQUIRED. We haven't done that before though, so there is that. cc @jdef

Reply:
Yeah if you define omitted keys as no-ops, then making the whole map optional makes no sense and the field should become required. I still think that there needs to be a way to unset previously-set mutable keys, which could be interpreted by the SP as reverting to some global default (defined elsewhere) or disabling some option (which was previously enabled).

Reply:
> I still think that there needs to be a way to unset previously-set mutable keys, which could be interpreted by the SP as reverting to some global default (defined elsewhere) or disabling some option (which was previously enabled).

I was thinking of adding a `boolean` flag to accomplish this if needed but I think a cleaner approach might be to add a new RPC call. Something like - `ControllerResetVolumeModification` which does what you say. 

But for now IMO, I am not sure if we need it and we can always come back to it and add it in future. This allows us to keep `ControllerModifyVolume` RPC call relatively simpler.

Reply:
So the things you suggest would also work, but they seem more complex. What's the downside of requiring the CO to pass all of the mutable options to each modify call and allowing the SP to interpret absent keys how it likes?

The point I'm trying to make here is that the CO implementation can be very simple and this detail doesn't matter (to the CO) because all the options are opaque anyways. The one thing I don't want to allow is updating a subset of the options in a single call, because I don't believe there is any benefit to allowing that.

Reply:
With my Kubernetes hat on, I think it should not matter what VAC was set to a PVC before, important is only the  current VAC.

In other words, if a volume is created with VAC `bronze`, then modified to `silver` and then to `gold`, it should have the same QoS parameters as if it was created with `gold` right away. Regardless if any parameter was / was not dropped from the VACs. Otherwise it will be very difficult to see what QoS parameters a volume actually has. And current proposal here will allow this behavior with little / no code on CO side.

Yes, it puts more responsibility on cluster admins, they must count with resetting a missing parameter to default, or even always file all parameters to all VACs.

Reply:
One idea: we could somehow clarify that `ControllerModifyVolume` should result in the same volume that would be created by `ControllerCreateVolume` with the same `parameters` as during original `ControllerCreateVolume` and new `mutable_parameters`. IMO that would be very clear to SP what to do, on the other hand they may need to store original `parameters` somewhere (in volume_context?), so they can return to them when corresponding key gets missing in `mutable_parameters`.

Reply:
That may not work for some SP's case, for example for GCP's hyperdisk-extreme, you can just specify the size of the disk, and a default size will be set for that disk even you did not specify one. But later you can actually modify the disk to have higher or lower iops as long as it is within the range.